### ISSUE TITLE
Items exclude from Discount

### ISSUE THREAD

**Issue by Koljan76** [frappe/erpnext#37697]
If we give Additional Discount Percentage in invoices, these discounts will be applied to all items. But some items may not be eligible for discounts, such as transportation costs, packaging, etc.
There must be a way to exclude such items from discounting.
One possible approach is to set a checkbox "Grant Discount" (similar to Grant Commission) in the item master data and to include this in the logic of the documents (offers, orders, invoices). If checkbox not selected, do not charge a discount for this item.

**Comment by Koljan76** [comment]
second thing. 
It is not possible to give a discount for items without a price list. Many companies do not use price lists, agree prices individually and have a discount

**Comment by bosue** [comment]
I agree there should be a basic use case without price lists, that should have an [ ] Allow discounts? switch.

Price lists should however add more fine-grained control. It should be possible to limit discounts on a per-Item basis either to a certain percentage or to a certain share of the margin.

Of course, the latter may be a separate issue deferred to a later point.

**Comment by skyasir** [comment]
For anyone who needs this before it lands in core: I've published a small app that does it, https://github.com/skyasir/erpnext_no_discount

- Adds a **No Discount Applicable** checkbox on Item (next to Grant Commission), copied to the item rows. A row can also be ticked by hand.
- Additional Discount skips flagged rows, and flagged rows are removed from the discount base: 10% on 30,000, with 10,000 flagged, gives a discount of 2,000, not 3,000.
- Works without a Price List, because it works from each row's `net_amount`.
- Covers every document with Additional Discount: Quotation, Sales Order, Delivery Note, Sales Invoice, POS Invoice, Supplier Quotation, Purchase Order, Purchase Receipt and Purchase Invoice. The flag carries over when one document is made from another.
- Works on v15 and v16 (uses `override_doctype_class`).

Known limit: with *Apply Discount On = Grand Total*, the flagged row's own tax still counts in the discount base. *Net Total* is exact.

**Comment by skyasir** [comment]
@umairsy @mihir-kandoi @DipenFrappe could you take a look? The app above implements this request (per-item exemption from Additional Discount, flag on Item, all selling and buying documents). If you'd like it in core instead, I'm happy to prepare it as a change to `taxes_and_totals` plus an Item field.
